kernel/unitycatalog/src/main/java/io/delta/kernel/unitycatalog/UnityCatalogUtils.java

```diff
@@ -0,0 +1,137 @@
+
+import static io.delta.kernel.commit.CatalogCommitterUtils.METASTORE_LAST_COMMIT_TIMESTAMP;
+import static io.delta.kernel.commit.CatalogCommitterUtils.METASTORE_LAST_UPDATE_VERSION;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.delta.kernel.commit.CatalogCommitterUtils;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.expressions.Column;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.clustering.ClusteringMetadataDomain;
+import io.delta.kernel.internal.util.ColumnMapping;
+import io.delta.kernel.internal.util.Tuple2;
+import io.delta.kernel.types.DataType;
+import io.delta.kernel.types.StructType;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class UnityCatalogUtils {
+  private UnityCatalogUtils() {}
+
+  private static final String UC_PROP_CLUSTERING_COLUMNS = "clusteringColumns";
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  /**
+   * Extract all properties that should be sent to Unity Catalog when creating a table (version 0).
+   *
+   * <p>This method extracts:
+   *
+   * <ul>
+   *   <li>All table properties from the metadata configuration
+   *   <li>Protocol-derived properties (e.g., delta.minReaderVersion=3, delta.feature.XXX=supported)
+   *   <li>UC-specific properties (delta.lastUpdateVersion, delta.lastCommitTimestamp)
+   *   <li>Clustering properties if a clustering domain metadata is present
+   * </ul>
+   *
+   * @param engine the engine to use for I/O operations (to retrieve the commit timestamp)
+   * @param postCreateSnapshot the snapshot after version 0 has been written
+   * @return a map of properties to send to Unity Catalog
+   * @throws IllegalArgumentException if the snapshot is not version 0
+   */
+  public static Map<String, String> getPropertiesForCreate(
+      Engine engine, SnapshotImpl postCreateSnapshot) {
+    if (postCreateSnapshot.getVersion() != 0) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Expected a snapshot at version 0, but got a snapshot at version %d",
+              postCreateSnapshot.getVersion()));
+    }
+
+    final Map<String, String> properties = new HashMap<>();
+
+    // Case 1: All table properties from metadata.configuration
+    properties.putAll(postCreateSnapshot.getTableProperties());
+
+    // Case 2: Protocol-derived properties
+    properties.putAll(
+        CatalogCommitterUtils.extractProtocolProperties(postCreateSnapshot.getProtocol()));
+
+    // Case 3: UC-specific properties
+    properties.put(METASTORE_LAST_UPDATE_VERSION, String.valueOf(postCreateSnapshot.getVersion()));
+    properties.put(
+        METASTORE_LAST_COMMIT_TIMESTAMP, String.valueOf(postCreateSnapshot.getTimestamp(engine)));
+
+    // Case 4: Clustering properties if present
+    postCreateSnapshot
```

Comment:
Since we are already using SnapshotImpl can we just use SnapshotImpl::getPhysicalClusteringColumns?

Reply:
Great questions! Will look into this 👍